### PR TITLE
Fix amp-forms broken and flakey tests.

### LINKED ISSUE #5775: AMP Form Flake
```
Chromium 51.0.2704 (Ubuntu 0.0.0) amp-form should allow rendering responses through templates FAILED
	AssertionError: expected null not to be null
	    at doAsserterAsyncAndAddThen (/home/travis/build/ampproject/amphtml/node_modules/chai-as-promised/lib/chai-as-promised.js:292:33)
	    at .<anonymous> (/home/travis/build/ampproject/amphtml/node_modules/chai-as-promised/lib/chai-as-promised.js:282:25)
	    at get (/home/travis/build/ampproject/amphtml/node_modules/chai/chai.js:5396:37)
	    at /tmp/c4798cd841690771cc0b2346e973d8e6.browserify:38735:43 <- /home/travis/build/ampproject/amphtml/extensions/amp-form/0.1/test/test-amp-form.js:451:42
```

### LINKED ISSUE #5789: amp-form user-valid tests failed
```
Chrome 54.0.2840 (Mac OS X 10.11.6) amp-form User Validity should manage valid/invalid on input/fieldset/form on submit FAILED
	AssertionError: expected 'user-valid valid' to include 'user-invalid'
	    at /var/folders/zy/2yz64_hx4y13mtfkjgyrw31h00bq__/T/702fd9c4794944464cdebe00d90ca033.browserify:39341:39 <- /Users/yuxichen/dev/amphtml/extensions/amp-form/0.1/test/test-amp-form.js:724:38
Chrome 54.0.2840 (Mac OS X 10.11.6) amp-form User Validity should manage valid/invalid on input user interaction FAILED
	AssertionError: expected 'user-valid valid' to include 'user-invalid'
	    at /var/folders/zy/2yz64_hx4y13mtfkjgyrw31h00bq__/T/702fd9c4794944464cdebe00d90ca033.browserify:39377:39 <- /Users/yuxichen/dev/amphtml/extensions/amp-form/0.1/test/test-amp-form.js:760:38

```

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #5775

### CODE REVIEW

File: src/log.js [hunk 1]
```diff
@@ -396,6 +420,7 @@ function createErrorVargs(var_args) {
  */
 export function rethrowAsync(var_args) {
   const error = createErrorVargs.apply(null, arguments);
+  recordAsyncErrorForTesting(error);
```
Comment (jridgewell) [inline src/log.js]: We really really really need to make sure this get's DCE.

Comment (mkhatib) [inline src/log.js]: @erwinmombay it should no - will let the next person who does a cut know to double check.

**Comment by aghassemi** [comment]
can mew haz this merged please? :cat: 

(I am changing Travis to run in proper Chrome with https://github.com/ampproject/amphtml/pull/5843 and this is always failing on Chrome 54 - or is a lot more flaky, I have not seen it pass yet)

**Comment by mkhatib** [comment]
Approve it and I'll merge it 😄

**Comment by aghassemi** [comment]
@mkhatib Then @jridgewell will (rightly) :hocho: me if that thing does not get DCE'd

File: src/log.js (hunk 1 above)
Comment (jridgewell) [inline src/log.js]: This is **not** DCE'd.

```js
function Ja(a){var b=Ha.apply(null,arguments);xa.push(b);setTimeout(function(){throw b;})}
```

File: src/log.js [hunk 2]
```diff
@@ -22,6 +22,30 @@ import {getModeObject} from './mode-object';
...
+}
+
+
+/**
+ * @param {!Error} error
+ * @visibleForTesting
+ */
+export function removeAsyncErrorForTesting(error) {
+  asyncErrorsThrownForTesting.splice(
+      asyncErrorsThrownForTesting.indexOf(error), 1);
```
Comment (jridgewell) [inline src/log.js]: You need to test that `#indexOf` doesn't return `-1`.

File: build-system/amp.extern.js [hunk 3]
```diff
@@ -35,6 +35,7 @@ window.draw3p;
 // AMP's globals
 window.AMP_TEST;
 window.AMP_TEST_IFRAME;
+window.onAsyncErrorForTesting;
```
Comment (jridgewell) [inline build-system/amp.extern.js]: Why is this exported?

File: extensions/amp-form/0.1/test/test-amp-form.js [hunk 4]
```diff
@@ -424,32 +426,39 @@ describe('amp-form', () => {
...
-        // Check that form has a rendered div with class .submit-error-message.
-        const renderedTemplate = form.querySelector('[i-amp-rendered]');
-        expect(renderedTemplate).to.not.be.null;
+      let asyncErrorRejecter;
+      const asyncErrorPromise = new Promise((unusedResolve, reject) => {
+        asyncErrorRejecter = reject;
+      });
+
+      // Catch expected rethrown async error.
+      window.onAsyncErrorForTesting(e => {
```
Comment (jridgewell) [inline extensions/amp-form/0.1/test/test-amp-form.js]: This doesn't seem like the right way to do this. Instead, we could export the module's `asyncErrorsThrownForTesting` and check that after a few seconds.

File: src/log.js (hunk 1 above)
Comment (cramforce) [inline src/log.js]: The fix is to guard this with a `getMode().` conditional.

File: src/log.js [hunk 5]
```diff
@@ -22,6 +22,30 @@ import {getModeObject} from './mode-object';
...
 /**
+ * @const {!Array<!Error>}
+ */
+const asyncErrorsThrownForTesting = [];
+
+
+/**
+ * @param {!Error} error
+ */
+function recordAsyncErrorForTesting(error) {
```
Comment (jridgewell) [inline src/log.js]: > provide API to force throwing async thrown errors in tests

That'll break the assumption of our source code. Just having a notification system we can check though would work.

**Comment by mkhatib** [comment]
I sent a separate PR #5861 to fix the broken tests (fieldset user-invalid stuff). For the flakey async errors I'll follow up on this PR.

File: src/log.js (hunk 1 above)
Comment (mkhatib) [inline src/log.js]: @jridgewell can you elaborate on "provide a function to swap it with another"?

File: build-system/amp.extern.js (hunk 3 above)
Comment (mkhatib) [inline build-system/amp.extern.js]: The usage inside `test-amp-form` complained about `function never defined` so I thought the right thing to do is this. Clearly it's not

File: extensions/amp-form/0.1/test/test-amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/test/test-amp-form.js]: > asyncErrorsThrownForTesting and check that after a few seconds

Yeah that should work too. Will play around with it.

**Comment by mkhatib** [comment]
Ok another go at this that only touches `form` tests. PTAL 👀 

If we'd like to come back to what @cramforce suggested to make sure async errors are thrown we can come back to that later. Just want to make sure form tests are not flakey/broken.

**Comment by cramforce** [comment]
Ship it. But please file an issue for async error solution.

**Comment by jridgewell** [comment]
Good enough for now.

**Comment by mkhatib** [comment]
Filed #5879 thanks for the reviews 😄

**Pull request merged**